File: hooks/block-unbounded-dependency.py
#!/usr/bin/env python3
"""PreToolUse hook (Edit|Write): block dependency versions without upper bounds."""

import json
import os
import re
import sys

_CARGO_WILDCARD_RE = re.compile(r"""['"](\s*\*\s*)['"]""")

_FILE_CHECKERS: dict[str, str] = {
    "pyproject.toml": "pyproject",
    "Cargo.toml": "cargo",
    "package.json": "package_json",
}


def _check_pyproject(content: str) -> bool:
    for line in content.splitlines():
        stripped = line.strip()
        if stripped.startswith("#"):
            continue
        if "requires-python" in stripped:
            continue
        if ">=" in stripped and "<" not in stripped and "~=" not in stripped:
            if re.search(r""">=\s*[\d]""", stripped):
                return True
    return False


def _check_cargo(content: str) -> bool:
    for line in content.splitlines():
        stripped = line.strip()
        if stripped.startswith("#"):
            continue
        if _CARGO_WILDCARD_RE.search(stripped):
            return True
        if ">=" in stripped and "<" not in stripped:
            if re.search(r""">=\s*[\d]""", stripped):
                return True
    return False


def _check_package_json(content: str) -> bool:
    for line in content.splitlines():
        stripped = line.strip()
        if ">=" in stripped and "<" not in stripped:
            if re.search(r""">=\s*[\d]""", stripped):
                return True
    return False


def main() -> None:
    data = json.load(sys.stdin)
    tool_name: str = data.get("tool_name", "")
    tool_input: dict[str, str] = data.get("tool_input", {})
    file_path: str = tool_input.get("file_path", "") or tool_input.get("path", "")

    if not file_path:
        return

    basename = os.path.basename(file_path)
    checker_key = _FILE_CHECKERS.get(basename)
    if not checker_key:
        return

    if tool_name == "Edit":
        content = tool_input.get("new_string", "")
    elif tool_name == "Write":
        content = tool_input.get("content", "")
    else:
        return

    if not content:
        return

    if checker_key == "pyproject":
        found = _check_pyproject(content)
    elif checker_key == "cargo":
        found = _check_cargo(content)
    elif checker_key == "package_json":
        found = _check_package_json(content)
    else:
        return

    if found:
        json.dump(
            {
                "decision": "block",
                "reason": (
                    "バージョン上限のない依存指定は禁止です（config: upper_bound_required = true）。\n"
                    ">= のみではなく、上限も指定してください（例: >=1.0,<2 / ~=1.0 / ^1.0）。"
                ),
            },
            sys.stdout,
        )


if __name__ == "__main__":
    main()

File: hooks-rs/src/hooks/block_unbounded_dependency.rs
use regex::Regex;

use crate::input::HookInput;
use crate::output::{block, pass};

pub fn run(input: &HookInput) {
    let file_path = input.tool_input.file_path_resolved();
    if file_path.is_empty() {
        return;
    }
    let basename = std::path::Path::new(file_path)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("");

    let content = match input.tool_name.as_str() {
        "Edit" => &input.tool_input.new_string,
        "Write" => &input.tool_input.content,
        _ => return,
    };
    if content.is_empty() {
        return;
    }

    let found = match basename {
        "pyproject.toml" => check_pyproject(content),
        "Cargo.toml" => check_cargo(content),
        "package.json" => check_package_json(content),
        _ => false,
    };

    if found {
        block(
            "バージョン上限のない依存指定は禁止です（config: upper_bound_required = true）。\n>= のみではなく、上限も指定してください（例: >=1.0,<2 / ~=1.0 / ^1.0）。",
        );
        return;
    }
    pass();
}

fn check_pyproject(content: &str) -> bool {
    let ge_re = Regex::new(r">=\s*[\d]").unwrap();
    for line in content.lines() {
        let stripped = line.trim();
        if stripped.starts_with('#') || stripped.contains("requires-python") {
            continue;
        }
        if stripped.contains(">=")
            && !stripped.contains('<')
            && !stripped.contains("~=")
            && ge_re.is_match(stripped)
        {
            return true;
        }
    }
    false
}

fn check_cargo(content: &str) -> bool {
    let wildcard_re = Regex::new(r#"['"](\s*\*\s*)['"]"#).unwrap();
    let ge_re = Regex::new(r">=\s*[\d]").unwrap();
    for line in content.lines() {
        let stripped = line.trim();
        if stripped.starts_with('#') {
            continue;
        }
        if wildcard_re.is_match(stripped) {
            return true;
        }
        if stripped.contains(">=") && !stripped.contains('<') && ge_re.is_match(stripped) {
            return true;
        }
    }
    false
}

fn check_package_json(content: &str) -> bool {
    let ge_re = Regex::new(r">=\s*[\d]").unwrap();
    for line in content.lines() {
        if line.contains(">=") && !line.contains('<') && ge_re.is_match(line) {
            return true;
        }
    }
    false
}
